src/components/TodoList/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,213 @@
+import React, { useEffect, useRef, useState } from 'react';
+import classNames from 'classnames';
+import * as todoService from '../../../api/todos';
+import { Todo } from '../../../types/Todo';
+import { useTodo } from '../../../api/useTodo';
+import { Error } from '../../../types/Error';
+
+type Props = {
+  todo: Todo;
+};
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
+  const {
+    setTodos,
+    setErrorMessage,
+    tempTodo,
+    toBeCleared,
+    setCleared,
+    toggled,
+    setToggled,
+  } = useTodo();
+
+  const [isTodoLoading, setIsTodoLoading] = useState(false);
+  const [isEditable, setIsEditable] = useState(false);
+  const [editedTitle, setEditedTitle] = useState(todo.title);
+
+  const editRef = useRef<HTMLInputElement>(null);
+  const editTimerRef = useRef<number | null>(null);
+
+  const handleDelete = (todoId: number, errorMessage = Error.Delete) => {
+    setIsTodoLoading(true);
+
+    todoService.deleteTodo(todoId)
+      .then(() => setTodos(currTodos => [...currTodos]
+        .filter(currTodo => currTodo.id !== todoId)))
+      .catch(() => {
+        setErrorMessage(errorMessage);
+      })
+      .finally(() => {
+        setIsTodoLoading(false);
+        setCleared([]);
+      });
+  };
+
+  const handleComplete = (
+    completedTodo: Todo,
+    errorMessage = Error.Update,
+  ) => {
+    setIsTodoLoading(true);
+
+    const updated = {
+      ...completedTodo,
+      completed: !completedTodo.completed,
+    };
+
+    todoService.updateTodo(todo.id, updated)
+      .then(updatedTodo => {
+        setTodos(currentTodo => {
+          const updateTodos = [...currentTodo];
+          const index = updateTodos
+            .findIndex(currTodo => currTodo.id === updatedTodo.id);
+
+          updateTodos.splice(index, 1, updatedTodo);
+
+          return updateTodos;
+        });
+      })
+      .catch(() => setErrorMessage(errorMessage))
+      .finally(() => {
+        setToggled([]);
+        setIsTodoLoading(false);
+      });
+  };
+
+  const handleEdit = (editedTodo: Todo) => {
```

Comment:
Try to avoid such nesting, use helper function 
![image](https://github.com/mate-academy/react_todo-app-with-api/assets/70009205/be432686-3be1-4283-87ca-6f5a4941ad2f)